File: Data Structures & Algorithms/subsets/submission-5.py
# class Solution:
#     def subsets(self, nums: List[int]) -> List[List[int]]:
#         self.res, self.track = [], []
#         self.backtrack(nums, 0)
#         return self.res
    
#     def backtrack(self, nums, start):
#         self.res.append(self.track[:])
#         n = len(nums)
#         if start > n:
#             return
        
#         for i in range(start, n):
#             self.track.append(nums[i])
#             self.backtrack(nums, i + 1)
#             self.track.pop()

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res, track = [], []
        def backtrack(i: int) -> None:
            if i == n:
                res.append(track[:])
                return
            
            backtrack(i+1)

            track.append(nums[i])
            backtrack(i+1)
            track.pop()

        backtrack(0) 
        return res
        

File: Data Structures & Algorithms/palindrome-partitioning/submission-5.java
// class Solution {
//     List<List<String>> res = new LinkedList<>();
//     List<String> track = new LinkedList<>();

//     public List<List<String>> partition(String s) {
//         dfs(s, 0);
//         return res;
//     }

//     private void dfs(String s, int start) {
//         if (start == s.length()) {
//             res.add(new ArrayList<>(track));
//             return;
//         }

//         for (int end = start; end < s.length(); end ++) {
//             if (!isPalindrome(s, start, end)) continue;
//             track.add(s.substring(start, end+1));
//             dfs(s, end+1);
//             track.remove(track.size() - 1);
//         }
//     }
//     private boolean isPalindrome(String s, int l, int r) {
//         while (l < r) {
//             if (s.charAt(l) != s.charAt(r)) {
//                 return false;
//             }

//             l ++;
//             r --;
//         }

//         return true;
//     }
// }


class Solution {
    List<List<String>> res = new LinkedList<>();
    List<String> track = new LinkedList<>();
    boolean[][] pal;

    public List<List<String>> partition(String s) {
        int n = s.length();
        pal = new boolean[n][n];

        for (int i = n-1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s.charAt(i) == s.charAt(j) && (j - i <= 2 || pal[i+1][j-1])) {
                    pal[i][j] = true;
                }
            }
        }

        dfs(s, 0);
        return res;
    }

    private void dfs(String s, int start) {
        if (start == s.length()) {
            res.add(new ArrayList<>(track));
            return;
        }

        for (int end = start; end < s.length(); end ++) {
            if (!pal[start][end]) continue;
            track.add(s.substring(start, end+1));
            dfs(s, end+1);
            track.remove(track.size() - 1);
        }
    }
}
